pkg/kv/kvserver/replica_write.go

```diff
@@ -884,3 +889,132 @@ func releaseMVCCStats(ms *enginepb.MVCCStats) {
+	// [), and we check that it doesn't overlap any rangeID local keys.
+	if span.EndKey == nil {
+		startRangeID, _, _, _, err := keys.DecodeRangeIDKey(span.Key)
+
+		if err != nil {
+			// If the start key is not a RangeID key, then this span cannot overlap
+			// with any rangeID local keys.
+			return nil //nolint:returnerrcheck
+		}
+
+		modifiedSpan := roachpb.Span{Key: span.Key, EndKey: span.Key.Next()}
+		if modifiedSpan.Overlaps(getRangeIDLocalSpan(startRangeID)) {
+			return errors.Errorf("overlapping an unreplicated rangeID span")
+		}
+
+		return nil
+	}
+
+	// (3) Check if the span is fully outside the unreplicated rangeID local span.
+	// In that case, we are certain that the given span doesn't overlap any
+	// unreplicated rangeID local keys.
+	if span.Key.Compare(keys.LocalRangePrefix) >= 0 ||
+		span.EndKey.Compare(keys.MakeRangeIDUnreplicatedPrefix(0)) <= 0 {
+		return nil
+	}
+
+	// (4) If the span fully overlaps the unreplicated rangeID local span, then
+	// we know that this span for sure overlaps some unreplicated rangeID keys.
+	if span.Key.Compare(keys.MakeRangeIDUnreplicatedPrefix(0)) <= 0 &&
+		span.EndKey.Compare(keys.LocalRangePrefix) >= 0 {
+		return errors.Errorf("overlapping an unreplicated rangeID span")
+	}
+
+	// (5) If the span partially overlaps the unreplicated range-local span.
+	if span.Key.Compare(keys.MakeRangeIDUnreplicatedPrefix(0)) < 0 &&
+		span.EndKey.Compare(keys.MakeRangeIDUnreplicatedPrefix(0)) > 0 {
+		return errors.Errorf("overlapping an unreplicated rangeID span")
+	}
+
+	if span.Key.Compare(keys.LocalRangePrefix) < 0 &&
+		span.EndKey.Compare(keys.LocalRangePrefix) > 0 {
+		return errors.Errorf("overlapping an unreplicated rangeID span")
+	}
+
+	// (6) At this point, we know that the given span is within the unreplicated
+	// rangeID local span. Check if this span could possibly intersect with any
+	// RangeID local keys.
+	startRangeID, _, _, _, err := keys.DecodeRangeIDKey(span.Key)
+	if err != nil {
+		return err
+	}
+
+	endRangeID, _, _, _, err := keys.DecodeRangeIDKey(span.EndKey)
+	if err != nil {
+		return err
+	}
+
+	if endRangeID-startRangeID > 1 {
+		return errors.Errorf("the span is accessing more than one rangeID local spans")
+	}
+
+	if span.Overlaps(getRangeIDLocalSpan(startRangeID)) ||
+		span.Overlaps(getRangeIDLocalSpan(endRangeID)) {
+		return errors.Errorf("overlapping an unreplicated rangeID span")
+	}
+
+	return nil
+}
+
+// overlapsStoreLocalKeys returns an error if the provided span overlaps
+// with any store local keys.
+func overlapsStoreLocalKeys(span roachpb.Span) error {
+	// If we receive a nil start key. It means that the span is effectively:
+	// {endKey.Prev(), endKey}. However, since we can't get the Prev() key, we
+	// set the startKey pessimistically to Prevish(roachpb.PrevishKeyLength).
+	if span.Key == nil {
+		span.Key = span.EndKey.Prevish(roachpb.PrevishKeyLength)
+	}
+
+	if span.EndKey == nil {
+		span.EndKey = span.Key.Next()
```

Comment:
This could allocate twice. Not a huge deal in tests, but still: how about doing the comparisons statelessly? Could we use the existing `overlaps` func? Also, we can probably avoid `Prevish`. 

For the `{Key: nil, EndKey: x}` case, essentially, the bad case is `start < x <= end` (as opposed to `start <= x < end` for `{Key: x, EndKey: nil}`).